File: webui/src/test/java/com/taskadapter/webui/action/MyTreeTest.java
package com.taskadapter.webui.action;

import com.taskadapter.core.PreviouslyCreatedTasksResolver;
import com.taskadapter.model.GTask;
import org.junit.Ignore;

import java.util.Arrays;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class MyTreeTest {

    private final GTask task0 = createTestTask(0L,
            createTestTask(1L), createTestTask(2L),
            createTestTask(3L,
                    createTestTask(4L),
                    createTestTask(5L,
                            createTestTask(6L))
            ),
            createTestTask(7L),
            createTestTask(8L,
                    createTestTask(9L)
            )
    );
    private final List<GTask> rootLevelTasks = List.of(task0);
    private final PreviouslyCreatedTasksResolver resolver = PreviouslyCreatedTasksResolver.empty;
    private final static List<GTask> expectedRootLevelTasks2 = List.of(
            createTestTask(0L, createTestTask(3L, createTestTask(5L, createTestTask(6L))))
    );

    private static GTask createTestTask(long id, GTask... children) {
        var task = new GTask();
        task.setId(id);
        task.setKey(id + "");
        task.setChildren(Arrays.asList(children));
        return task;
    }

    @Ignore
    public void allTasksAreSelectedByDefault() {
        var myTree = new MyTree(resolver, rootLevelTasks, "some location");
        var selectedGTaskList = myTree.getSelectedRootLevelTasks();
        assertThat(selectedGTaskList).isNotNull();
        assertThat(selectedGTaskList).containsAll(rootLevelTasks);
    }

    private final List<GTask> expectedRootLevelTasks1 = List.of(createTestTask(0L, createTestTask(1L), createTestTask(2L),
            createTestTask(7L), createTestTask(8L, createTestTask(9L)))
    );

    @Ignore
    public void allChildrenAreUnselectedWhenParentIsUnselected() {
        var myTree = new MyTree(resolver, rootLevelTasks, "some location");
        // deselect parent #3
//    myTree.tree.getContainerProperty(new TaskId(3L, "3"), MyTree.ACTION_PROPERTY).getValue.asInstanceOf[Checkbox].setValue(false)
        var selectedGTaskList = myTree.getSelectedRootLevelTasks();
        assertThat(selectedGTaskList).isNotNull();
        assertThat(selectedGTaskList).containsAll(expectedRootLevelTasks1);
    }

    @Ignore
    public void allParentsUpTheHierarchyMustBeSelectedIfAtLeastOneChildIsSelected() {
        var myTree = new MyTree(resolver, rootLevelTasks, "some location");
        // deselect parent #0
//    myTree.tree.getContainerProperty(new TaskId(0L, "0"), MyTree.ACTION_PROPERTY).getValue.asInstanceOf[Checkbox].setValue(false)
        // select child #6
//    myTree.tree.getContainerProperty(new TaskId(6L, "6"), MyTree.ACTION_PROPERTY).getValue.asInstanceOf[Checkbox].setValue(true)
        var selectedGTaskList = myTree.getSelectedRootLevelTasks();
        assertThat(selectedGTaskList).isNotNull();
        assertThat(selectedGTaskList).containsAll(expectedRootLevelTasks2);
    }
}

File: webui/src/test/scala/com/taskadapter/webui/action/MyTreeTest.scala
package com.taskadapter.webui.action

import com.taskadapter.connector.definition.TaskId
import com.taskadapter.core.PreviouslyCreatedTasksResolver
import com.taskadapter.model.GTask
import com.vaadin.ui.CheckBox
import org.junit.Assert
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.scalatest.{FunSpec, Matchers}

import scala.collection.JavaConverters._

@RunWith(classOf[JUnitRunner])
class MyTreeTest extends FunSpec with Matchers {

  val task0 = createTestTask(0l,
    createTestTask(1l), createTestTask(2l),
    createTestTask(3l,
      createTestTask(4l),
      createTestTask(5l,
        createTestTask(6l))
    ),
    createTestTask(7l),
    createTestTask(8l,
      createTestTask(9l)
    )
  )
  var rootLevelTasks = Seq(task0)

  val resolver = PreviouslyCreatedTasksResolver.empty

  private def createTestTask(id: Long, children: GTask*) = {
    val task = new GTask
    task.setId(id)
    task.setKey(id + "")
    task.setChildren(children.asJava)
    task
  }

  it("all tasks are selected by default") {
    val myTree = new MyTree(resolver, rootLevelTasks.asJava, "some location")
    val selectedGTaskList = myTree.getSelectedRootLevelTasks
    Assert.assertNotNull(selectedGTaskList)
    selectedGTaskList.asScala shouldBe rootLevelTasks
  }

  val expectedRootLevelTasks1 = Seq(createTestTask(0l, createTestTask(1l), createTestTask(2l),
    createTestTask(7l), createTestTask(8l, createTestTask(9l)))
  )

  it("all children are unselected when parent is unselected") {
    val myTree = new MyTree(resolver, rootLevelTasks.asJava, "some location")
    // deselect parent #3
    myTree.tree.getContainerProperty(TaskId(3l, "3"), MyTree.ACTION_PROPERTY).getValue.asInstanceOf[CheckBox].setValue(false)
    val selectedGTaskList = myTree.getSelectedRootLevelTasks
    Assert.assertNotNull(selectedGTaskList)
    selectedGTaskList.asScala shouldBe expectedRootLevelTasks1
  }

  it("all parents up the hierarchy must be selected if at least one child is selected") {
    val myTree = new MyTree(resolver, rootLevelTasks.asJava, "some location")
    // deselect parent #0
    myTree.tree.getContainerProperty(TaskId(0l, "0"), MyTree.ACTION_PROPERTY).getValue.asInstanceOf[CheckBox].setValue(false)
    // select child #6
    myTree.tree.getContainerProperty(TaskId(6l, "6"), MyTree.ACTION_PROPERTY).getValue.asInstanceOf[CheckBox].setValue(true)
    val selectedGTaskList = myTree.getSelectedRootLevelTasks
    Assert.assertNotNull(selectedGTaskList)
    selectedGTaskList.asScala.toList shouldBe expectedRootLevelTasks2.toList
  }

  val expectedRootLevelTasks2 = Seq(
    createTestTask(0l, createTestTask(3l, createTestTask(5l, createTestTask(6l))))
  )
}
